File: 01507. Reformat Date/solution.cpp
/**
 * Problem: 1507. Reformat Date
 *
 * Difficulty: Easy
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution {
public:
  string reformatDate(string date) {
    // Split date into day, month, and year components via string stream
    string dayWithSuffix, monthAbbrev, yearString;
    stringstream stream(date);

    // Parse day with suffix, month abbreviation, and year
    stream >> dayWithSuffix >> monthAbbrev >> yearString;

    // Extract day number by removing ordinal suffix (last 2 chars) and pad to 2
    // digits
    string dayNumber = dayWithSuffix.substr(0, dayWithSuffix.size() - 2);
    string dayString = dayNumber.size() == 1 ? "0" + dayNumber : dayNumber;

    // Map month abbreviation to two-digit month number
    unordered_map<string, string> monthMap = {
        {"Jan", "01"}, {"Feb", "02"}, {"Mar", "03"}, {"Apr", "04"},
        {"May", "05"}, {"Jun", "06"}, {"Jul", "07"}, {"Aug", "08"},
        {"Sep", "09"}, {"Oct", "10"}, {"Nov", "11"}, {"Dec", "12"}};

    // Look up month number from abbreviation
    string monthString = monthMap[monthAbbrev];

    // Return formatted date as YYYY-MM-DD
    return yearString + "-" + monthString + "-" + dayString;
  }
};

File: 01507. Reformat Date/solution.cs
/**
 * Problem: 1507. Reformat Date
 *
 * Difficulty: Easy
 *
 * Language: C#
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

public class Solution
{
  public string ReformatDate(string date)
  {
    // Extract month abbreviation via range and map to two-digit number using switch expression
    string month = date[^8..^5] switch
    {
      "Jan" => "01",
      "Feb" => "02",
      "Mar" => "03",
      "Apr" => "04",
      "May" => "05",
      "Jun" => "06",
      "Jul" => "07",
      "Aug" => "08",
      "Sep" => "09",
      "Oct" => "10",
      "Nov" => "11",
      _ => "12"
    };

    // Return formatted date as YYYY-MM-DD, padding single-digit day with leading zero
    return char.IsDigit(date[1])
        ? $"{date[^4..]}-{month}-{date[..2]}"
        : $"{date[^4..]}-{month}-0{date[..1]}";
  }
}
